### ISSUE TITLE
Merge API does not work in MySql

### ISSUE THREAD

**Issue by hardhub** [linq2db/linq2db#1596]
```
Exception message:
You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'MERGE INTO `Language` Target
USING
(
	VALUES
	(0x08D68D024AAA54E8F8E0E388D57' at line 1
Stack trace:

   at MySql.Data.MySqlClient.MySqlStream.ReadPacket()
   at MySql.Data.MySqlClient.NativeDriver.GetResult(Int32& affectedRow, Int64& insertedId)
   at MySql.Data.MySqlClient.Driver.NextResult(Int32 statementId, Boolean force)
   at MySql.Data.MySqlClient.MySqlDataReader.NextResult()
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader(CommandBehavior behavior)
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery()
   at LinqToDB.Data.DataConnection.ExecuteNonQuery()
   at LinqToDB.Data.CommandInfo.Execute()
   at LinqToDB.DataProvider.BasicMerge.Execute(DataConnection dataConnection)
   at LinqToDB.Data.DataConnectionExtensions.Merge[T](DataConnection dataConnection, IEnumerable`1 source, String tableName, String databaseName, String schemaName)

```
### Environment details
linq2db version: *2.6.0,2.6.4*
Database Server: *MySql 8.0*
Database Provider: *MySql.Data 8.0.15*
Operating system: *E.g. Windows 7 / CentOS 7*
Framework version: *.NET Framework 4.6.2 / Mono (latest)*

**Comment by hardhub** [comment]
It looks like MySql should use `ON DUPLICATE KEY UPDATE`

But it uses something different.

@[sdanyliv](/sdanyliv) do you have comments here? :)

I afraid I have big issue in case of using of MySql.

**Comment by sdanyliv** [comment]
Can you provide your `Merge` query?

**Comment by hardhub** [comment]
Not sure it makes sense but it looks this way:

```C#
public async Task<long> Update<P>(Array<P> planes, int batch)
            where P : class, IPlane
        {
            if (planes.Count == 1)
                return await Storage.GetTask(() => (long)this.Update(planes.Single()));

            if (planes.Count >= 2)
            {
                Func<Array<P>, long> action = array =>
                {
                    long total = 0L;

                    using (TransactionScope transaction = new TransactionScope())
                    {
                        int iterations = array.Count / batch + (array.Count % batch > 0 ? 1 : 0);

                        for (int index = 0; index < iterations; index++)
                        {
                            IEnumerable<P> scope = array.Skip(index * batch).Take(batch);

                            total += this.Merge(scope);
                        }

                        transaction.Complete();
                    }

                    return total;
                };

                return await Storage.GetTask(() => action(planes));
            }

            return 0L;
        }
```

Where is `P` is any class generated by T4 template from table.

**Comment by sdanyliv** [comment]
@hardhub, looks like we have do nothing with MySql in this area. As i know, MySql do not support Merge.

**Comment by hardhub** [comment]
Does it make any sense?

https://stackoverflow.com/questions/2469457/insert-on-duplicate-key-update-with-where/13947397#13947397

**Comment by sdanyliv** [comment]
Yes, but, i have checked code and nothing done in this area for MySql in linq2db.

**Comment by hardhub** [comment]
Can you give me some links so I will try to implement something?
It is a bit complex to understand how exactly it should be written if you do not work with Linq2db on regular basis.

**Comment by sdanyliv** [comment]
You have to implement `Merge` and `MergeAsync` members for `MySqlDataProvider`
https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/DataProvider/IDataProvider.cs#L72

**Comment by hardhub** [comment]
Do you have any implementation in code except SQL Server?

I see there is some RDBMS supporting that: https://vladmihalcea.com/how-do-upsert-and-merge-work-in-oracle-sql-server-postgresql-and-mysql/

**Comment by hardhub** [comment]
I guess all is around BasicMerge, even OracleMerge uses BuildUsing2

**Comment by MaceWindu** [comment]
duplicates #1480

**Comment by MaceWindu** [comment]
`BasicMerge` is legacy SQL Server-only implementation with a lot of issues/limitations.

Current merge implemented by BasicMergeBuilder and it's descendands for concrete providers and will be substituted by #1512 in v3

**Comment by hardhub** [comment]
What is the state of 3.0 version and what is working branch for it?
If I want to compile and preview it, is it possible for the moment?

**Comment by MaceWindu** [comment]
Right now 3.0 branch (`release.3.0`) doesn't have anything new merged compared to 2.7 (`master`) - new features are still in development/blocked by merge refactoring.
Status of branches could be viewed here https://github.com/linq2db/linq2db/wiki/Releases-and-Roadmap#release-300-unreleased - all merged features are documented in corresponding release section

**Comment by MaceWindu** [comment]
Gonna close it as duplicate of #1480
